eth/consensus/clique/clique.py

```diff
@@ -0,0 +1,189 @@
+from typing import Sequence, Iterable
+
+from eth.db.chain import ChainDB
+from eth_typing import (
+    Address,
+    Hash32,
+)
+from eth_utils import (
+    decode_hex,
+    encode_hex,
+    to_tuple,
+    ValidationError,
+)
+
+from eth.rlp.headers import BlockHeader
+
+from eth.typing import VMConfiguration
+from eth.vm.chain_context import ChainContext
+from eth.vm.execution_context import (
+    ExecutionContext,
+)
+# from eth.vm.forks.byzantium.headers import (
+#     create_header_from_parent,
+#     configure_header,
+# )
+
+from .datatypes import (
+    Snapshot,
+)
+from .header_cache import HeaderCache
+from .snapshot_manager import SnapshotManager
+from ._utils import (
+    get_signer_address,
+    humanize_hashes,
+    validate_header_integrity,
+)
+
+# https://github.com/ethereum/EIPs/issues/225
+# https://github.com/ethereum/go-ethereum/blob/0ac9bbba6cd259bc9895e786121cdcb6e7c5d9d2/consensus/clique/clique.go
+
+
+
+def create_execution_context(self,
+                             prev_hashes: Iterable[Hash32],
+                             chain_context: ChainContext,
+                             import_block: BlockHeader=None) -> ExecutionContext:
+
+    # In Clique consensus, the tx fee goes to the signers so for the scope of the
+    # Execution context we set the coinbase to the signer if possible.
+
+    if import_block is not None:
+        try:
+            coinbase = get_signer_address(import_block.header)
+        except Exception:
+            coinbase = self.coinbase
+    else:
+        coinbase = self.coinbase
+
+    return ExecutionContext(
+        coinbase=coinbase,
+        timestamp=self.timestamp,
+        block_number=self.block_number,
+        difficulty=self.difficulty,
+        gas_limit=self.gas_limit,
+        prev_hashes=prev_hashes,
+        chain_id=chain_context.chain_id,
+    )
+
+BlockHeader.create_execution_context = create_execution_context
+
+class CliqueConsensus:
+
+    logger = logging.getLogger('eth.consensus.clique.CliqueConsensus')
+
+    def __init__(self, base_db, epoch_length: int = 30000) -> None:
+        if base_db is None:
+            raise Exception("Meh")
+        self._current_signers = []
+        self._epoch_length = epoch_length
+        self._chain_db = ChainDB(base_db)
+        self._header_cache = HeaderCache(self._chain_db)
+        self._snapshot_manager = SnapshotManager(
+            self._chain_db,
+            self._header_cache,
+            self._epoch_length,
+        )
+
+    @to_tuple
+    def amend_vm_configuration(self, config: VMConfiguration) -> VMConfiguration:
+        for pair in config:
+            block_number, vm = pair
+            vm_class = vm.configure(
+                block_extra_data_max_bytes=65535,
+                validate_seal=lambda header: self.validate_seal(header),
+                configure_header=self.configure_header
+            )
+
+            yield block_number, vm_class
+
+    def construct_wrong_turn_message(self,
```

Comment:
I think "out_of_turn" is more intuitive verbage for this (wrong turn in english conveys more like you took a wrong turn in your car).

Reply:
The Clique protocol has a concept of blocks that are coming "in turn" or "out of turn" and both are valid (just getting a different difficulty assigned). This function is constructing an error message for when difficulty and "turness" don't match up. So using "out of turn" might be confusing here as "out of turn" isn't necessarily wrong in Clique consensus.

Would `construct_turn_error_message` feel natural as well?